### ISSUE TITLE
Change the scope of algebras to protected

### ISSUE THREAD

**Issue by MateuszKubuszok** [endpoints4s/endpoints4s#553]
When building endpoints using DSL the exact scope of all algebras and their values, methods etc is not really relevant, as long as you can access it while building things.

However, the moment you interpret algebra into something specific: client, server, etc, you can find that all these internals leaking outside are confusing to newcomers, juniors etc. It is difficult to explain to them e.g. why HTTP client besides methods for calling server has also tons of things related to processing status code, serialization, raw queries, etc shown when they try to list available methods from in their IDE.

For that I suggest that (unless there are some reasons to not do this) change the scope of all methods (and possibly some types) to `protected`. This would make them still available to whoever is building the endpoint, make it possible to interpret algebra to a specific implementation, but the user who uses the generated interpretation would only see the methods we exposed to them ourselves, making it more user friendly. If we needed to expose some of the internals we could still do it ourselves, but as a conscious decision and not as a byproduct of a leaking abstraction.

Ideally, user of the interpreter should see only methods that were implemented by API trait as public, and see only the types that are returned by endpoints interpretation. Probably it won't be always possible, but I believe that for majority of cases it is doable and beneficial.

**Comment by julienrf** [comment]
That’s an interesting point. Another problem with the current design is that having everything public makes the compilation slow (I think).

However, it is sometimes useful to be able to use algebra methods from the outside. We do it [here](https://github.com/julienrf/endpoints/blob/3b28c7280b2e96edd4fe71f144ed84509ac92b6f/openapi/openapi/src/main/scala/endpoints/openapi/JsonSchemas.scala#L230) for example to use the `ujson.JsonSchemas` interpreter to build codecs for us while we build our `openapi` interpreter.

**Comment by MateuszKubuszok** [comment]
From what I see in your example code is using [this value](https://github.com/julienrf/endpoints/blob/3b28c7280b2e96edd4fe71f144ed84509ac92b6f/openapi/openapi/src/main/scala/endpoints/openapi/JsonSchemas.scala#L29)? I think it could be made `protected`? I think it could be possible to use some sed to make all definitions `protected` and see where it would break things. Probably in many places it would be fine (which would improve things immediately) and the rest could be considered by case.

**Comment by julienrf** [comment]
@MateuszKubuszok, the definition you linked could definitely be made protected. However, the members we access _via_ this definition could not be made protected. And we basically access all the members of the `JsonSchemas` algebra.

**Comment by julienrf** [comment]
After thinking more about it, I think we can’t change the visibility of algebra operations to `protected` (for the reasons given above).

Maybe a possible workaround could be to use composition instead of inheritance:

~~~ scala
trait MyEndpoints { // *no* “extends algebra.Endpoints”
  protected val api: algebra.Endpoints
  import api._

  val someEndpoint = endpoint(get(path / "foo"), ok(textResponse))
}

object MyServer extends MyEndpoints {
  protected lazy val api: server.Endpoints = new ... // Instantiate your preferred server interpreter
}

MyServer.someEndpoint(()) // Only `someEndpoint` is visible
~~~

(see also #102)

**Comment by MateuszKubuszok** [comment]
I played around with making things protected, and indeed there were some problems. But yes, probably if - instead of one godlike class at the interpreter site - there were a few separate, smaller classes combined by responsibility (endpoints themselves, JSON format, etc) it could be easier to achieve.

It could be a bit more works, but still, I believe it could pay off in a long run. I've noticed that right now if I e.g. try to use circe, then depending on whether I want to derive Swagger (which means that I derive codecs from schema) on if I need to use mu own codecs (which let me use codecs directly, but makes swagger a bit harder) I need to resolve some cake-pattern-like challenges. Splitting things into smaller, independent algebras could make things easier to compose, I think. So it would be 2 improvements at once. But, yeah, that would be bigger change than I though.
